### DISCUSSION TITLE
Finding Values

### DISCUSSION THREAD

**Discussion by VinylVault** [joalla/discogs_client#32] [Support] [answered]
Hi Folks, I'm trying to find 2 values in the release data, and struggling like mad, I've been through every `dir()` that I can expand. so if anyone can tell me where to find these two keys, that would be amazing :D

Date added to collection - please note, this is not the date added to discogs, which is already available.
Folder ID for a release.

I can find the ID's of all my folders, and add the releases to my local database by folder, so finding the folder ID isn't a game breaker, unlike the date added to collection.  

Hope someone can give me some advice

**Comment by alifhughes** [comment, marked as answer]
Hey @VinylVault. Good question! So the collection Folder ID of the release can be found [here](https://github.com/joalla/discogs_client/blob/master/discogs_client/models.py#L614) (code example below). The date added to collection is not actually available in the client.. _yet_. I could see that it is available in [the API response](https://www.discogs.com/developers/#page:user-collection,header:user-collection-collection-items-by-release) so I am going to create a PR to add it to this client 🙂 Thanks for pointing this out! Code example for both here:
```python
>>> import discogs_client
>>> d = discogs_client.Client('ExampleApplication/0.1')
>>> d.set_consumer_key('...', '...')
>>> d.get_authorize_url()[2]
>>> .....
>>> d.get_access_token('.....')
>>> me = d.identity()
>>> me.collection_folders
[<CollectionFolder 0 'All'>, <CollectionFolder 1 'Uncategorized'>]
>>> collection = me.collection_folders[0]
>>> collection.releases[0]
<CollectionItemInstance 7580596 'Amen EP Remixes'>
>>> collection.releases[0].folder_id
1

# NOT AVAILABLE YET, WILL BE IN SUBSEQUENT PR
>>> collection.releases[0].date_added
datetime.datetime(2016, 8, 31, 16, 11, 23, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=61200)))
```

**Reply by alifhughes** [reply]
[PR here](https://github.com/joalla/discogs_client/pull/33) to add `date_added` to collection item. Once this is merged and released I will mark this as answered if that is all 🙂

**Reply by VinylVault** [reply]
@alifhughes can these two items please be brought forward into the standard release information rather than having to dig them out? .. accessible as 
```
d = discogs_client.Client('somename/0.4', user_token="**hidden**")

me = d.identity()

collection = me.collection_folders

cursor = connection.cursor()

for shelf in collection:
    if shelf.id in [0]
        for album in shelf.releases:
            release = album.release
            release.folder_id
            release.date_added_to_collection
```
I think this accessibility would be far more beneficial to more users, the `date_added` would need to be changed to `date_added_to_collection`  as `date_added` is already in use for the date the release was added to discogs

**Reply by alifhughes** [reply]
Hi @VinylVault. If I understand what you're asking, I don't know if it would be beneficial to add this information at a release level rather than a collection level. Think about it in the contrary, what if you simply wanted to get a release item without having to authenticate, without it being connected to a collection? In the example you give you're looping through items in a collection, which is exactly what the model `CollectionItemInstance` is for, it is an item in your collection. If we were to add it at a release level and you simply wanted to do something like:
```python
>>> import discogs_client
>>> d = discogs_client.Client('ExampleApplication/0.1')
# Note: no authentication which is a perfectly normal path as specified in the documentation.
>>> release = d.release(1234)
>>> release.date_added_to_collection
?????
```
In this context, it would not make sense because we do not have an authenticated user, simply someone who would like to get information about a release/artist/anything. Does that make sense? Thanks

**Reply by VinylVault** [reply]
I can see your thoughts, is it possible for it to be available **both** ways? seeing as a 'CollectionItemInstance' object is not iterable. Also, if a user isn't authenticated, personal collection information isn't available, so neither field would be populated for a request that isn't made by an authenticated user, or am i wrong?

Just in case I'm going about things in an unexpected way, how would you pull every its in the collection into a local database? obviously, I've missed great chunks of my code out above, for things like artists, record labels and tracks

**Reply by alifhughes** [reply]
> is it possible for it to be available both ways?

I personally don't think this is a good solution, mainly because I think it could become quite complex quickly with all the different scenarios of being authenticated and how that would make different properties of the different models appear or not. But most of all, I think it wouldn't be good solution because it doesn't represent the actual API response, which our client at the moment does. If you see the response [here](https://www.discogs.com/developers#page:user-collection,header:user-collection-collection-items-by-release), the way the properties work on each model in the client is the same as it is in the API response.

> seeing as a 'CollectionItemInstance' object is not iterable.

A `CollectionItemInstance` is one collection item inside a `Collection`, and a `Collection` is iterable. You can see an example below of me iterating through all my releases in a collection which I think is what you're trying to achieve?

```python
import discogs_client

d = discogs_client.Client('ExampleApplication/0.1', user_token="...")
me = d.identity()

# One of my collection folders
collection = me.collection_folders[0]

for collection_item in collection.releases:

    collection_folder_id = collection_item.folder_id

    # To be added in PR: 33
    # date_added = collection_item.date_added

    release = collection_item.release
    # And then can get anything from the release as normal
    release_artists = release.artists
    print(collection_folder_id, release, release_artists)
```

Does that help? If it does, please let me know and I'll mark this as answered. Thanks

**Reply by VinylVault** [reply]
i guess, time to refactor my code base

**Reply by VinylVault** [reply]
@alifhughes when will it push to pypi etc?

**Reply by alifhughes** [reply]
The `date_added` will be pushed once it has a review, I'll chase to get the reviews today - everything else in the code snippet works as is

**Reply by JOJ0** [reply]
@VinylVault If you need a feature to move on with your own coding, you can always temporarily install a version of discogs_client from a contributors feature branch temporarily as I suggested recently in PRs #33 and #34.
HTH :-)

**Reply by alifhughes** [reply]
This has been released now @VinylVault in version 2.3.10 - link to download it via PyPI [here](https://pypi.org/project/python3-discogs-client/2.3.10/)
